File: Recursion/E-509. Fibonacci Number/E-509. Fibonacci Number.js
/**
 * @param {number} n
 * @return {number}
 */
var fib = function (n) {
  return calcFib(n, {});
};

const calcFib = (n, cache) => {
  if (n === 0 || n == 1) return n;
  if (n in cache) return cache[n];
  const result = calcFib(n - 1, cache) + calcFib(n - 2, cache);
  cache[n] = result;
  return result;
};

File: Recursion/E-509. Fibonacci Number/E-509. Fibonacci Number.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int fib(int n) {
        Map<Integer, Integer> cache = new HashMap<>();
        return Solution.getFib(n, cache);
    }

    public static int getFib(int n, Map<Integer, Integer> cache) {
        if (cache.containsKey(n)) {
            return cache.get(n);
        }

        if (n <= 1) {
            return n;
        }

        return Solution.getFib(n - 2, cache) + Solution.getFib(n - 1, cache);

    }
}
